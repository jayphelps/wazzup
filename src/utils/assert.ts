export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
