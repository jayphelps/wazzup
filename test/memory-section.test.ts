import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Memory Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds a memory', () => {
    module.addMemory(1, 1, []);

    assertModule(module, `
      | (module
      |   (memory (;0;) 1 1))
    `);
  });

  test('adds a memory with a single datum', () => {
    const data = [
      new ast.Data('hello world', new ast.Const(0, wasm.Opcode.I32_CONST))
    ];

    module.addMemory(2, 4, data);

    assertModule(module, `
      | (module
      |   (memory (;0;) 2 4)
      |   (data (i32.const 0) "hello world"))
    `);
  });

  test('adds a memory with a multiple data', () => {
    const data = [
      new ast.Data('hello', new ast.Const(0, wasm.Opcode.I32_CONST)),
      new ast.Data('world', new ast.Const(50, wasm.Opcode.I32_CONST))
    ];

    module.addMemory(2, 4, data);

    assertModule(module, `
      | (module
      |   (memory (;0;) 2 4)
      |   (data (i32.const 0) \"hello\")
      |   (data (i32.const 50) \"world\"))
    `);
  });
});
