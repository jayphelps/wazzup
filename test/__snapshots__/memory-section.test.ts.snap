// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Memory Section adds a memory 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [
    Memory {
      "externalKind": 2,
      "initialSize": 1,
      "maxSize": 1,
      "segments": Array [],
    },
  ],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Memory Section adds a memory 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  5,
  4,
  1,
  1,
  1,
  1,
]
`;

exports[`Memory Section adds a memory with a multiple data 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [
    Memory {
      "externalKind": 2,
      "initialSize": 2,
      "maxSize": 4,
      "segments": Array [
        Data {
          "offset": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 0,
          },
          "value": "hello",
        },
        Data {
          "offset": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 50,
          },
          "value": "world",
        },
      ],
    },
  ],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Memory Section adds a memory with a multiple data 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  5,
  4,
  1,
  1,
  2,
  4,
  11,
  21,
  2,
  0,
  65,
  0,
  11,
  5,
  104,
  101,
  108,
  108,
  111,
  0,
  65,
  50,
  11,
  5,
  119,
  111,
  114,
  108,
  100,
]
`;

exports[`Memory Section adds a memory with a single datum 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [
    Memory {
      "externalKind": 2,
      "initialSize": 2,
      "maxSize": 4,
      "segments": Array [
        Data {
          "offset": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 0,
          },
          "value": "hello world",
        },
      ],
    },
  ],
  "tables": Array [],
  "types": Array [],
  "version": 1,
}
`;

exports[`Memory Section adds a memory with a single datum 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  5,
  4,
  1,
  1,
  2,
  4,
  11,
  17,
  1,
  0,
  65,
  0,
  11,
  11,
  104,
  101,
  108,
  108,
  111,
  32,
  119,
  111,
  114,
  108,
  100,
]
`;
